src/report_reader.cpp

```diff
@@ -470,7 +470,7 @@ DataFrame<T> ReportReader<T>::Population::get(const nonstd::optional<Selection>&
                 data_ptr[offset] = buffer[gid_start];
             } else {  // Elements report
                 uint64_t gid_end = range.second - min;
-                std::copy(&buffer[gid_start], &buffer[gid_end], &data_ptr[offset]);
+                std::copy(buffer.begin() + gid_start, buffer.begin() + gid_end, &data_ptr[offset]);
```

Comment:
Hmmmm... Do you mind to clarify why the change? I don't mind, but if we do change this, we should also change the code in line 470 and we must verify that the performance of the copies remains the same (e.g., this operation is executed thousands of times in thousands of timesteps, it matters!).

On a side note, perhaps using something like `std::next(buffer.begin(), gid_start)` as well.

Reply:
Turning on the `_GLIBCXX_ASSERTIONS` makes doing the subscript on buffer look like `&buffer[buffer.size()]`, which would be an out of bounds error and thus undefined behavior.  AFAIK, this is ok in this *specific* case as long as `std::copy` doesn't deref the pointer.

The change means we use the `std::copy` overload that takes iterators, and is "safer", because there is no chance there is an OOB deref, as long as `gid_end` never is larger than the size of the array.

The ASM will look the same from both of them, I'd expect, at a reasonable optimization level.

Reply:
Thanks @mgeplf for clarifying. My point was that accessing the array with `[]` in `std::copy` might avoid boundary checks or additional code, but if you feel that it should be the same, then no complains at all.

I'd still prefer the use of `std::next(buffer.begin(), gid_start)` and `std::next(buffer.begin(), gid_end)` with iterators, although I don't mind leaving the code as it is with the `+` operator.

Reply:
> but if you feel that it should be the same, then no complains at all.

Had a quick look on godbolt to confirm (which I never take as conclusive), but it seems that compiler sees through this:
https://godbolt.org/z/8dez5nEM3

> I'd still prefer the use of std::next(buffer.begin(), gid_start)

Ok, I'll make those changes, and another one to remove the pointer usage altogether.

Reply:
I've made the changes; any/all reviews welcome.

Reply:
All tests pass in my Fedora build with the newest changes :+1: